test/jdk/java/lang/StringBuffer/ECoreIndexOf.java

```diff
@@ -0,0 +1,1370 @@
+      foo += indexOfKernel(sb, c, 0);
+      foo += indexOfKernel(sb, c16);
+      foo += indexOfKernel(sb, c16, 0);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////
+    ///////////////////////////////////////////////////////////////////////////////////
+
+    String[] decorators = {"", " (same char)"};
+    Charset[] charSets = {StandardCharsets.ISO_8859_1, StandardCharsets.UTF_16};
+    boolean[] truefalse = {true, false};
+
+    titles.put(StandardCharsets.ISO_8859_1, "L");
+    titles.put(StandardCharsets.UTF_16, "U");
+
+    for (int xxy = 0; xxy < 2; xxy++) { // Run at least twice to ensure stub called
+
+      for (int i = 0; i < 128; i++) {
+        haystack[i] = (char) i;
+      }
+
+      haystack_16[0] = '\u0000'; // (char) (23 + 256);
+      for (int i = 1; i < 128; i++) {
+        haystack_16[i] = (char) (i);
+      }
+
+      simpleTest();
+      compareIndexOfLastIndexOf();
+      compareStringStringBuffer();
+      StringIndexof();
+      StringIndexofChar();
+      StringIndexofHuge();
+
+      for (String decorator : decorators) {
+        for (Charset csHaystack : charSets) {
+          for (Charset csNeedle : charSets) {
+            System.out.println("Testing " + titles.get(csHaystack) + titles.get(csNeedle) + decorator);
+            for (boolean useOffset : truefalse) {
+              for (boolean useBuffer : truefalse) {
+                exhaustive(useOffset, useBuffer, csHaystack, csNeedle);
+              }
+            }
+          }
+        }
+
+        for (int i = 0; i < 128; i++) {
+          haystack[i] = (char) 'a';
+        }
+
+        for (int i = 0; i < 128; i++) {
+          haystack_16[i] = (char) ('a' + 256);
+        }
+      }
+    }
+
+    System.out.println(testName + " complete.");
+
+    if (failure)
+      throw new RuntimeException("One or more failures.");
+  }
+
+  private static void report(String testName, int failCount) {
+    System.err.println(testName + ": " +
+        (failCount == 0 ? "Passed" : "Failed(" + failCount + ")"));
+    if (failCount > 0)
+      failure = true;
+  }
+
+  private static String generateTestString(int min, int max) {
+    StringBuffer aNewString = new StringBuffer(120);
+    int aNewLength = getRandomIndex(min, max);
+    for (int y = 0; y < aNewLength; y++) {
+      int achar = generator.nextInt(30) + 30;
+      char test = (char) (achar);
+      aNewString.append(test);
+    }
+    return aNewString.toString();
+  }
+
+  private static String makeRndString(boolean isUtf16, int length) {
+    StringBuilder sb = new StringBuilder(length);
+    if (length > 0) {
+      sb.append(isUtf16 ? '\u2026' : 'b'); // ...
+
+      for (int i = 1; i < length - 1; i++) {
+        sb.append((char) ('b' + rng.nextInt(26)));
+      }
+
+      sb.append(rng.nextInt(3) >= 1 ? 'a' : 'b');// 66.6% of time 'a' is in string
+    }
+    return sb.toString();
+  }
+
+  private static int indexOfKernel(String haystack, String needle) {
```

Comment:
Is the intention of kernels not to be inlined so that it would be part of separate compilation?

If so, you probably want to annotate it with `@CompilerControl(CompilerControl.Mode.DONT_INLINE)`

i.e. https://github.com/openjdk/jmh/blob/master/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_16_CompilerControl.java

Reply:
Fixed.

Reply:
CompilerControl is unavailable here.  Added a runtime option instead.